File: algorithms/1422.maximum-score-after-splitting-a-string/solution.go
func maxScore(s string) int {
	max := 0
	size := len(s)
	scores := make([]int, size-1)

	countZero := 0
	countOne := 0

	for left := 0; left < size-1; left++ {
		if s[left] == '0' {
			countZero += 1
		}
		scores[left] += countZero
	}

	for right := size - 1; right > 0; right-- {
		if s[right] == '1' {
			countOne += 1
		}
		scores[right-1] += countOne

		if scores[right-1] > max {
			max = scores[right-1]
		}
	}

	return max
}

File: algorithms/1422.maximum-score-after-splitting-a-string/solution.rs
impl Solution {
    pub fn max_score(s: String) -> i32 {
        let len = s.len();
        let bits: Vec<char> = s.chars().collect();
        let mut scores = vec![0; len - 1];

        let mut count_zero = 0;
        for left in 0..len - 1 {
            if bits[left] == '0' {
                count_zero += 1;
            }
            scores[left] += count_zero;
        }

        let mut count_one = 0;
        for right in (0..len - 1).rev() {
            if bits[right + 1] == '1' {
                count_one += 1;
            }
            scores[right] += count_one;
        }

        *scores.iter().max().unwrap() as i32
    }
}
